testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java

```diff
@@ -1044,37 +1049,195 @@ public void testUserSessionRefreshAndIdleRememberMe() throws Exception {
+            tokenResponse = oauth.doAccessTokenRequest(code, "password");
+            assertTrue("Invalid ExpiresIn", 0 < tokenResponse.getRefreshExpiresIn() && tokenResponse.getRefreshExpiresIn() <= 1000);
+            events.expectCodeToToken(loginEvent.getDetails().get(Details.CODE_ID), sessionId).assertEvent();
+
+            clientSessionId = getClientSessionUuid(sessionId, loginEvent.getClientId());
+            assertTrue("User session expired", testingClient.testing().cache(InfinispanConnectionProvider.USER_SESSION_CACHE_NAME).contains(sessionId));
+            assertTrue("Client session expired", testingClient.testing().cache(InfinispanConnectionProvider.CLIENT_SESSION_CACHE_NAME).containsUuid(clientSessionId));
+
+            setTimeOffset(3700);
+            tokenResponse = oauth.doRefreshTokenRequest(tokenResponse.getRefreshToken(), "password");
+            assertEquals(400, tokenResponse.getStatusCode());
+            assertNull(tokenResponse.getAccessToken());
+            assertNull(tokenResponse.getRefreshToken());
+            events.expect(EventType.REFRESH_TOKEN).error(Errors.INVALID_TOKEN).user((String) null).assertEvent();
+            assertFalse("User session not expired", testingClient.testing().cache(InfinispanConnectionProvider.USER_SESSION_CACHE_NAME).contains(sessionId));
+            assertFalse("Client session not expired", testingClient.testing().cache(InfinispanConnectionProvider.CLIENT_SESSION_CACHE_NAME).containsUuid(clientSessionId));
         } finally {
+            getTestingClient().testing().revertTestingInfinispanTimeService();
             RealmManager.realm(realmResource).ssoSessionMaxLifespan(maxLifespan);
+            RealmManager.realm(realmResource).ssoSessionIdleTimeout(idleTimeout);
+            RealmManager.realm(realmResource).clientSessionMaxLifespan(clientMaxLifespan);
+            RealmManager.realm(realmResource).clientSessionIdleTimeout(clientIdleTimeout);
+            events.clear();
+            resetTimeOffset();
+        }
+    }
+
+    @Test
+    public void refreshTokenUserClientMaxLifespanGreaterThanSession() throws Exception {
+        RealmResource realmResource = adminClient.realm("test");
+        Integer maxLifespan = realmResource.toRepresentation().getSsoSessionMaxLifespan();
+        Integer idleTimeout = realmResource.toRepresentation().getSsoSessionIdleTimeout();
+        Integer clientMaxLifespan = realmResource.toRepresentation().getClientSessionMaxLifespan();
+        Integer clientIdleTimeout = realmResource.toRepresentation().getClientSessionIdleTimeout();
+        getTestingClient().testing().setTestingInfinispanTimeService();
+        try {
+            RealmManager.realm(realmResource).ssoSessionMaxLifespan(3600);
+            RealmManager.realm(realmResource).ssoSessionIdleTimeout(7200);
+            RealmManager.realm(realmResource).clientSessionMaxLifespan(5000);
+            RealmManager.realm(realmResource).clientSessionIdleTimeout(7200);
```

Comment:
This effectively means four very slow calls to update realm parameters upon setting and another four upon resetting in the `finally` block. Please consider replacling all these calls in all methods affected in this PR with `RealmAttributeUpdater`, and removing the changes in `RealmManager`.

For an example usage, see https://github.com/keycloak/keycloak/blob/f5ebe675a845f26d338d7ccc64d8781fb6445597/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java#L808-L812

Reply:
Yeah! Good point! Done in the added methods!